### DISCUSSION TITLE
How to handle large amount of data and concurrent use

### DISCUSSION THREAD

**Discussion by mermicode** [xyflow/xyflow#1682] [Q&A]
Hey all!

We love this project! and working on a project we have potential 100's of nodes and edges. We're using a DB to update any changes to the diagram in order to display the diagram to multiple users and increase collaboration.

As we're doing that, we're hitting a few problems with scaling and rendering and wanted to see if anyone has come across this or that can share some ideas!

1. The data is updated very frequently, which is slowing down the performance of the application. Does anyone have any ideas of being able to feed react-flow with continuous data without it loosing performance?
2. 'Locking' a node - our requirement needs to allow for a user to work on a node with no disruption that are happening else where (like editing the node text) - how can you 'fence off' a specific node? We're finding this very difficult as no hashmap is used, we have to loop through the list of nodes and edges each time there is an update or a user selects a node.
3. In general, how do you find the node in the array in an efficient way without using a hashmap?

Thank you all in advance! :)

**Comment by xplato** [comment]
> Does anyone have any ideas of being able to feed react-flow with continuous data without it loosing performance?

One recommendation I can make here, which you and your team may already be doing, is to keep the `elements` state as close to where it's needed as possible.

I'd add a few `useEffect` calls with an empty dependency array to see how often your container component(s) is (are) being rendered. If, for the sake of argument, you have your state in a provider a few (1-10) components up in the component tree, you may find that your entire view is being re-rendered unnecessarily.

Moreover, this sounds like the perfect time to [`memo` your components](https://reactjs.org/docs/react-api.html#reactmemo). If you're using a custom hook, you can use [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) instead.

I know these may be some elementary suggestions, but I hope they point you in the right direction.
